kb/testing/testing-ci.md

```diff
@@ -0,0 +1,107 @@
+
+The `win32-mingw` component runs the following bat script:
+```
+cmake . -G "MinGW Makefiles" -DCMAKE_C_COMPILER="gcc"
+mingw32-make
+mingw32-make test
+ctest -VV
+programs\test\selftest.exe
+```
+
+The `win32_msvc12_32` component runs the following bat script:
+```
+call "C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\vcvarsall.bat"
+cmake . -G "Visual Studio 12"
+MSBuild ALL_BUILD.vcxproj
+programs\test\Debug\selftest.exe
+```
+The `win32-msvc12_64` component is identical except that it runs `cmake . -G "Visual Studio 12 Win64"`.
+
+## Tooling
+
+### Tooling for all.sh
+
+To run `tests/scripts/all.sh`, you need at least the following tools:
+
+* basic Unix/POSIX shell utilities. Currently any modern `sh` should do, but a dependency on `bash` may be added in 2020.
+* Git. `all.sh` requires a Git checkout. The following files may be restored to their checked-in version: `include/mbedtls/config.h`, `**/Makefile`.
+
+`all.sh` is composed of many components, some of which have additional requirements. Most components require the following tools:
+
+* `gcc` or `clang` (depending on the component) for native builds.
+* GNU make or CMake (depending on the component).
+* Perl 5.
+* Python ≥3..
+* The commands `openssl`, `gnutls-cli` and `gnutls-serv` must be present in the `$PATH` to run any test component (otherwise `all.sh` refuses to start). If you don't have them, you can still run some test components that don't do any interoperability testing with `env OPENSSL=false GNUTLS_CLI=false GNUTLS_SERV=false tests/scripts/all.sh …`.
+
+### Running `all.sh` on Ubuntu
+
+Our reference Linux platform is Ubuntu 16.04 x86 64-bit. More recent Ubuntu versions should also work.
+
+On Ubuntu 16.04, the following packaged tools are known to cause problems with `all.sh`:
+
+* The `gnutls-bin` package on Ubuntu 16.04 does not work with our test scripts because Ubuntu has removed some obsolete features.
+
+On Ubuntu 20.04, the following packaged tools are known to cause problems with `all.sh`:
+
+* `all.sh check_doxygen_warnings` fails. Doxygen 1.8.11 and 1.8.13 are known to work.
+
+On Ubuntu, for the 32-bit builds, install the following packages:
+```
+apt-get install gcc-multilib libc6-dev-i386 linux-libc-dev:i386 libc6:386
+```
+Instead of `gcc-multilib`, which conflicts with cross-compilers for non-Intel architectures, a versioned package (e.g. `gcc5-multilib`) will do, but you need to create an additional symbolic link:
+```
+ln -s x86_64-linux-gnu/asm /usr/include/asm
+```
+
+### Notes about networking
+
+The tests attempt to establish an IPv6 connection to the local host. This is optional, and will be skipped if IPv6 is not available. However, if IPv6 is possible but IPv6 connections to localhost are blocked by a firewall, the test will fail.
+
+### Docker environment
+
+A Dockerfile is provided in [tests/docker](https://github.com/Mbed-TLS/mbedtls/tree/development/tests/docker).
```

Comment:
I wonder if we should keep some of the warning from the original along the lines of "This Dockerfile is not identical to the one used in the CI (XXX is it?) and is infrequently updated" or something like that.

How does this one compare to the one at (for example) https://github.com/Mbed-TLS/mbedtls-test/blob/master/resources/docker_files/ubuntu-18.04/Dockerfile ? (and 16.04, 20.04, 22.04)

Would we be better off removing this one and referring to the other(s)?

Reply:
I don't know enough about how we use docker to know how best to document this. If you want to suggest something better I'm happy to take it.

Reply:
https://github.com/Mbed-TLS/mbedtls-test/blob/master/resources/docker_files/ubuntu-18.04/Dockerfile is not infrequently updated, it's unmaintained and to be removed. Part of the work of removing it is to remove references to it in various documentation, so please do not reference it in any new document.

Reply:
I've removed the section; I'll leave it to someone who knows more about our docker usage to add an updated version in a separate PR. (@bensze01  ?)

Reply:
@gilles-peskine-arm The various Dockerfiles in https://github.com/Mbed-TLS/mbedtls-test/blob/master/resources/docker_files/ubuntu-??.04/Dockerfile were updated several times in 2023, including by you in September. Are we really removing those? I've found them useful

If you mean the one at https://github.com/Mbed-TLS/mbedtls/tree/development/tests/docker referenced in this text, then yes, it seems unmaintained, and we shouldn't refer to it

Reply:
Oops, right, I copied the wrong link. I meant the one referenced in this text, which was an external contribution back when the ones in `mbedtls-test` weren't public. We have no plan to do anything other than routine maintenance to the dockerfiles in `mbedtls-test`.